File: heaps/sort_characters_by_frequency.py
import heapq

class Solution:
    def frequencySort(self, s: str) -> str:
        freq_count = {}
        for char in s:
            freq_count[char] = freq_count.get(char, 0) + 1

        heap = [(-count, char) for char, count in freq_count.items()]
        heapq.heapify(heap)

        result = []
        while heap:
            count, char = heapq.heappop(heap)
            result.append(char * -count)


        return ''.join(result)

File: heaps/sort_characters_by_frequency.rb
require 'algorithms'
# @param {String} s
# @return {String}
def frequency_sort(s)
    #   create a hashmap with letters as keys and values as frequency
    freq_count = Hash.new(0)
    s.each_char { |char| freq_count[char] += 1 }
    
    #   create the max heap from the hashmap result will be an array
    heap = Containers::MaxHeap.new
    freq_count.each do |char, count|
      heap.push([count, char])
    end

    result = ""

    #   iterate the hashmap to collect values from the keys
    
    until heap.empty?
      count, char = heap.pop

      #   populate new str
      result << char * count
    end

    result
end